### PR TITLE
Remove the erase regions MIR transform

### PR DESCRIPTION
**Pull request by matthewjasper**
We now ensure that MIR never contains unerased regions in the first place.

### CODE REVIEW

File: src/librustc_mir_build/build/mod.rs [hunk 1]
```diff
@@ -209,7 +218,8 @@ fn liberated_closure_env_ty(
     };
 
     let closure_env_ty = tcx.closure_env_ty(closure_def_id, closure_substs).unwrap();
-    tcx.liberate_late_bound_regions(closure_def_id, &closure_env_ty)
+    let re_erased = tcx.lifetimes.re_erased;
+    tcx.replace_late_bound_regions(&closure_env_ty, |_| re_erased).0
```
Comment (eddyb) [inline src/librustc_mir_build/build/mod.rs]: Doesn't `erase_late_bound_regions` exist? I vaguely recall codegen using it.

File: src/librustc_mir_build/build/mod.rs [hunk 2]
```diff
@@ -128,12 +127,8 @@ fn mir_build(tcx: TyCtxt<'_>, def_id: DefId) -> BodyAndCache<'_> {
                 let ty = if fn_sig.c_variadic && index == fn_sig.inputs().len() {
                     let va_list_did =
                         tcx.require_lang_item(lang_items::VaListTypeLangItem, Some(arg.span));
-                    let region = tcx.mk_region(ty::ReScope(region::Scope {
-                        id: body.value.hir_id.local_id,
-                        data: region::ScopeData::CallSite,
-                    }));
```
Comment (eddyb) [inline src/librustc_mir_build/build/mod.rs]: Can we see a simplification of `ty::RegionKind` some time soon? :grin:

Comment (matthewjasper) [reply src/librustc_mir_build/build/mod.rs]: I'm working on it.

**Review by eddyb — APPROVED** [review]
r=me after switching to `erase_late_bound_regions` (if it exists)
cc @nikomatsakis

**Review by nikomatsakis — APPROVED** [review]

**Pull request merged**